File: hackers-revenge-server/app/models/running_program.rb
class RunningProgram
  MAX_STACK_SIZE = 16

  # Prevent enormous numbers from being concocted (e.g., via DUPE ; ADD ; JUMP -2).
  MIN_ARG_VALUE = -1_000_000
  MAX_ARG_VALUE = 1_000_000

  attr_reader :program_num # 1 or 2
  attr_reader :memory # array of MemoryCell
  attr_reader :stack # array of integers
  attr_reader :ip # current instruction pointer (in 0..255)

  # After calling run_next_instruction, these tell you what ip and stack were before it was called.
  attr_reader :old_stack
  attr_reader :old_ip

  # After calling run_next_instruction, these tell you which opcode (and arg) was just ran.
  attr_reader :ran_opcode
  attr_reader :ran_argument

  # After calling run_next_instruction, these tell you which addresses were just read/written to.
  attr_reader :read_addr_first
  attr_reader :read_addr_last
  attr_reader :wrotes # array of hashes containing :addr, :opcode, and :argument; yes, I know wrotes isn't a word

  def initialize(program_num, memory, start_ip, instructions)
    fail "RunningProgram: No instructions" if instructions.size.zero?
    fail "RunningProgram: Too many instructions" if instructions.size > ::Program::MAX_INSTRUCTIONS

    @program_num = program_num.to_i
    @memory = memory
    @stack = []
    @ip = start_ip % 256

    @old_stack = stack.dup
    @old_ip = ip.dup
    @ran_opcode = @ran_argument = nil
    @read_addr_first = @read_addr_last = @wrotes = nil

    load_to_memory(instructions)
  end

  # Return value is the running status (see status enum in Cycle).
  def run_next_instruction
    @old_stack = stack.dup
    @old_ip = ip.dup
    @ran_opcode = @ran_argument = nil
    @read_addr_first = @read_addr_last = @wrotes = nil

    inst = memory[ip]
    return ::Cycle::STATUS_DIED_EXECUTING_NIL if inst.nil? || inst.opcode.nil?

    @ran_opcode = inst.opcode
    @ran_argument = inst.argument

    return ::Cycle::STATUS_DIED_ARG_OVERFLOW if arg < MIN_ARG_VALUE || arg > MAX_ARG_VALUE

    case @ran_opcode
    when "NOOP" then run_noop
    when "PUSH" then run_push
    when "DROP" then run_drop
    when "DUPE" then run_dupe
    when "DUPE2" then run_dupe2
    when "SWAP" then run_swap
    when "INC" then run_inc
    when "ADD" then run_add
    when "SUBTRACT" then run_subtract
    when "DIVIDE" then run_divide
    when "MULTIPLY" then run_multiply
    when "GT" then run_gt
    when "LT" then run_lt
    when "NEGATE" then run_negate
    when "JUMP" then run_jump
    when "JUMPZ" then run_jumpz
    when "JUMPG" then run_jumpg
    when "HCF" then run_hcf(inst)
    when "USCAN" then run_uscan
    when "SCAN" then run_scan
    when "COPY" then run_copy
    when "ICOPY" then run_copy(:indirect => true)
    when "SETARG" then run_setarg
    when "PUSHARG" then run_pusharg
    when "INCARG" then run_incarg
    when "DECSKIP" then run_decskip
    else ::Cycle::STATUS_DIED_INVALID_OPCODE
    end
  end

private

  def arg
    @ran_argument.to_i
  end

  def inc_ip
    @ip = (@ip + 1) % 256
    ::Cycle::STATUS_RUNNING
  end

  def load_to_memory(instructions)
    pos = ip
    instructions.each do |inst|
      memory[pos] = ::MemoryCell.new(:opcode => inst.opcode, :argument => inst.argument, :program_num => program_num)
      pos = (pos + 1) % 256
    end
  end

  def run_add
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.size < 2
    value = stack.pop + stack.pop
    return ::Cycle::STATUS_DIED_ARG_OVERFLOW if value < MIN_ARG_VALUE || value > MAX_ARG_VALUE
    stack << (value % 256)
    inc_ip
  end

  def run_subtract
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.size < 2
    arg1 = stack.pop
    arg2 = stack.pop
    value = arg2 - arg1
    return ::Cycle::STATUS_DIED_ARG_OVERFLOW if value < MIN_ARG_VALUE || value > MAX_ARG_VALUE
    stack << (value % 256)
    inc_ip
  end

  def run_divide
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.empty?
    arg2 = arg
    if arg2 == 0
      pass
    else
      arg1 = stack.pop
      value = arg1 / arg2
      return ::Cycle::STATUS_DIED_ARG_OVERFLOW if value < MIN_ARG_VALUE || value > MAX_ARG_VALUE
      stack << (value % 256)
    end
    inc_ip
  end

  def run_multiply
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.empty?
    arg2 = arg
    if arg2 == 0
      pass
    else
      arg1 = stack.pop
      value = arg1 * arg2
      return ::Cycle::STATUS_DIED_ARG_OVERFLOW if value < MIN_ARG_VALUE || value > MAX_ARG_VALUE
      stack << (value % 256)
    end
    inc_ip
  end

  def run_gt
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.empty?
    stack << (stack.pop > arg ? 1 : 0)
    inc_ip
  end

  def run_lt
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.empty?
    stack << (stack.pop < arg ? 1 : 0)
    inc_ip
  end

  def run_copy(indirect: false)
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.empty?
    src_pos = (ip + arg) % 256
    dst_pos = (ip + stack.pop) % 256
    dst_pos = (dst_pos + memory[dst_pos]&.argument.to_i) % 256 if indirect
    explosion = memory[src_pos]&.hcf? && memory[dst_pos]&.hcf?
    memory[dst_pos] = memory[src_pos]

    @read_addr_first = @read_addr_last = src_pos
    wrote = { :addr => dst_pos }
    unless memory[dst_pos].nil?
      wrote[:opcode] = memory[dst_pos].opcode
      wrote[:argument] = memory[dst_pos].argument
    end
    @wrotes = [wrote]

    if explosion
      dst_pos_before = (dst_pos + 255) % 256
      # spread fire to all neighboring fire
      while memory[dst_pos_before]&.hcf?
        dst_pos_before += 255
        dst_pos_before %= 256
      end

      dst_pos_after = (dst_pos + 1) % 256
      # spread fire to all neighboring fire
      while memory[dst_pos_after]&.hcf?
        dst_pos_after += 1
        dst_pos_after %= 256
      end
      memory[dst_pos_before] = memory[dst_pos_after] = memory[dst_pos]

      @wrotes = [wrote.dup, wrote.dup, wrote.dup]
      @wrotes[0][:addr] = dst_pos_before
      @wrotes[2][:addr] = dst_pos_after
    end

    inc_ip
  end

  def run_decskip
    pos = (ip + arg) % 256

    if memory[pos].nil?
      memory[pos] = ::MemoryCell.new_no_opcode(:argument => -1, :program_num => program_num)
    else
      memory[pos].inc_argument(-1)
      memory[pos].program_num = program_num
    end

    @wrotes = [{ :addr => pos, :argument => memory[pos].argument }]

    inc_ip if memory[pos].argument.zero?
    inc_ip
  end

  def run_drop
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.empty?
    stack.pop
    inc_ip
  end

  def run_dupe
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.empty?
    return ::Cycle::STATUS_DIED_STACK_OVERFLOW if stack.size >= MAX_STACK_SIZE
    stack << stack[-1]
    inc_ip
  end

  def run_dupe2
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.empty?
    return ::Cycle::STATUS_DIED_STACK_OVERFLOW if stack.size >= MAX_STACK_SIZE
    stack << stack[-1]
    return ::Cycle::STATUS_DIED_STACK_OVERFLOW if stack.size >= MAX_STACK_SIZE
    stack << stack[-1]
    inc_ip
  end

  def run_hcf(inst)
    if inst.program_num == program_num
      ::Cycle::STATUS_DIED_EXECUTING_OWN_HCF
    else
      ::Cycle::STATUS_DIED_EXECUTING_ENEMY_HCF
    end
  end

  def run_inc
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.empty?
    stack[-1] += arg
    inc_ip
  end

  def run_jump
    return ::Cycle::STATUS_DIED_JUMP_0 if arg.zero?
    @ip = (@ip + arg) % 256
    ::Cycle::STATUS_RUNNING
  end

  def run_jumpg
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.empty?
    if stack.pop.positive?
      run_jump
    else
      inc_ip
    end
  end

  def run_jumpz
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.empty?
    if stack.pop.zero?
      run_jump
    else
      inc_ip
    end
  end

  def run_negate
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.empty?
    stack[-1] *= -1
    inc_ip
  end

  def run_noop
    inc_ip
  end

  def run_push
    return ::Cycle::STATUS_DIED_STACK_OVERFLOW if stack.size >= MAX_STACK_SIZE
    stack << arg
    inc_ip
  end

  def run_setarg
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.empty?
    value = stack.pop
    pos = (ip + arg) % 256

    if memory[pos].nil?
      memory[pos] = ::MemoryCell.new_no_opcode(:argument => value, :program_num => program_num)
    else
      memory[pos].argument = value
      memory[pos].program_num = program_num
    end

    @wrotes = [{ :addr => pos, :argument => value }]

    inc_ip
  end

  def run_pusharg
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.empty?
    pos = (ip + stack.pop) % 256
    value = memory[pos]&.argument.to_i
    stack << value
    inc_ip
  end

  def run_incarg
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.empty?
    pos = (ip + stack.pop) % 256

    if memory[pos].nil?
      memory[pos] = ::MemoryCell.new_no_opcode(:argument => arg, :program_num => program_num)
    else
      memory[pos].inc_argument(arg)
      memory[pos].program_num = program_num
    end

    @wrotes = [{ :addr => pos, :argument => memory[pos].argument }]

    inc_ip
  end

  def run_uscan
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.empty?
    v = stack.pop
    pos = (ip + v) % 256
    count = arg
    inc_val = count
    inst_found = nil

    @read_addr_first = pos unless count.zero?
    while count != 0
      inst_found = memory[pos]
      break unless inst_found.nil? || inst_found.opcode.nil?

      if count.positive?
        pos = (pos + 1) % 256 # ++pos
        count -= 1
      else
        pos = (pos + 255) % 256 # --pos
        count += 1
      end
    end
    @read_addr_last = pos if @read_addr_first.present?

    # Reverse first and last when scan was backwards.
    if arg.negative?
      @read_addr_first, @read_addr_last = @read_addr_last, @read_addr_first
    end

    if inst_found.nil? || inst_found.opcode.nil? || inst_found.hcf?
      v += inc_val
    else
      v -= 1
    end
    stack << v
    inc_ip
  end

  def run_scan
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.empty?
    pos = (ip + stack.pop) % 256
    count = arg
    inst_found = nil

    @read_addr_first = pos unless count.zero?
    while count != 0
      inst_found = memory[pos]
      break unless inst_found.nil? || inst_found.opcode.nil?

      if count.positive?
        pos = (pos + 1) % 256 # ++pos
        count -= 1
      else
        pos = (pos + 255) % 256 # --pos
        count += 1
      end
    end
    @read_addr_last = pos if @read_addr_first.present?

    # Reverse first and last when scan was backwards.
    if arg.negative?
      @read_addr_first, @read_addr_last = @read_addr_last, @read_addr_first
    end

    if inst_found.nil? || inst_found.opcode.nil?
      stack << 0
    elsif inst_found.hcf?
      stack << -1
    else
      stack << 1
    end

    inc_ip
  end

  def run_swap
    return ::Cycle::STATUS_DIED_STACK_UNDERFLOW if stack.size < 2
    tmp = stack[-1]
    stack[-1] = stack[-2]
    stack[-2] = tmp
    inc_ip
  end
end

File: hackers-revenge-go/internal/engine/running_program.go
package engine

import (
	"fmt"
	"os"
	"strconv"
	"strings"

	"github.com/xyloguy/hackers-revenge/hackers-revenge-go/internal/models"
)

const (
	MaxStackSize = 16
	MinArgValue  = -1000000
	MaxArgValue  = 1000000
	MemorySize   = 256
)

// RunningProgram represents a program being executed
type RunningProgram struct {
	ProgramNum int           // 1 or 2
	Memory     []*MemoryCell // shared memory (256 cells)
	Stack      []int         // program's stack
	IP         int           // instruction pointer (0-255)

	// State tracking for the last instruction
	OldStack      []int
	OldIP         int
	RanOpcode     string
	RanArgument   *int
	ReadAddrFirst *int
	ReadAddrLast  *int
	Wrotes        []map[string]interface{}
}

// NewRunningProgram creates a new running program
func NewRunningProgram(programNum int, memory []*MemoryCell, startIP int, instructions []models.Instruction) (*RunningProgram, error) {
	if len(instructions) == 0 {
		return nil, fmt.Errorf("no instructions provided")
	}
	if len(instructions) > models.MaxInstructions {
		return nil, fmt.Errorf("too many instructions")
	}

	rp := &RunningProgram{
		ProgramNum: programNum,
		Memory:     memory,
		Stack:      make([]int, 0, MaxStackSize),
		IP:         startIP % 256,
		OldStack:   make([]int, 0),
		OldIP:      startIP % 256,
	}

	// Load instructions into memory
	if err := rp.loadToMemory(instructions); err != nil {
		return nil, err
	}

	return rp, nil
}

// loadToMemory loads the program's instructions into shared memory
func (rp *RunningProgram) loadToMemory(instructions []models.Instruction) error {
	pos := rp.IP
	for _, inst := range instructions {
		cell, err := NewMemoryCell(inst.Opcode, inst.Argument, rp.ProgramNum)
		if err != nil {
			return err
		}
		rp.Memory[pos] = cell
		pos = (pos + 1) % 256
	}
	return nil
}

// RunNextInstruction executes the next instruction and returns the status
func (rp *RunningProgram) RunNextInstruction() int {
	// Save state
	rp.OldStack = make([]int, len(rp.Stack))
	copy(rp.OldStack, rp.Stack)
	rp.OldIP = rp.IP
	rp.RanOpcode = ""
	rp.RanArgument = nil
	rp.ReadAddrFirst = nil
	rp.ReadAddrLast = nil
	rp.Wrotes = nil

	// Get instruction at current IP
	inst := rp.Memory[rp.IP]
	if inst == nil || inst.Opcode == "" {
		return models.StatusDiedExecutingNil
	}

	rp.RanOpcode = inst.Opcode
	rp.RanArgument = inst.Argument

	// Check for argument overflow
	arg := rp.arg()
	if arg < MinArgValue || arg > MaxArgValue {
		return models.StatusDiedArgOverflow
	}

	// Execute the instruction
	switch rp.RanOpcode {
	case "NOOP":
		return rp.runNoop()
	case "PUSH":
		return rp.runPush()
	case "DROP":
		return rp.runDrop()
	case "DUPE":
		return rp.runDupe()
	case "DUPE2":
		return rp.runDupe2()
	case "SWAP":
		return rp.runSwap()
	case "INC":
		return rp.runInc()
	case "ADD":
		return rp.runAdd()
	case "SUBTRACT":
		return rp.runSubtract()
	case "DIVIDE":
		return rp.runDivide()
	case "MULTIPLY":
		return rp.runMultiply()
	case "GT":
		return rp.runGT()
	case "LT":
		return rp.runLT()
	case "NEGATE":
		return rp.runNegate()
	case "JUMP":
		return rp.runJump()
	case "JUMPZ":
		return rp.runJumpZ()
	case "JUMPG":
		return rp.runJumpG()
	case "HCF":
		return rp.runHCF(inst)
	case "SCAN":
		return rp.runScan()
	case "COPY":
		return rp.runCopy(false, 0)
	case "ICOPY":
		return rp.runCopy(true, 0)
	//case "COPY2":
	//	return rp.runCopy(false, 1)
	case "SETARG":
		return rp.runSetArg()
	case "PUSHARG":
		return rp.runPushArg()
	case "INCARG":
		return rp.runIncArg()
	case "DECSKIP":
		return rp.runDecSkip()
	default:
		return models.StatusDiedInvalidOpcode
	}
}

// Helper functions

func (rp *RunningProgram) arg() int {
	if rp.RanArgument == nil {
		return 0
	}
	return *rp.RanArgument
}

func (rp *RunningProgram) incIP() int {
	rp.IP = (rp.IP + 1) % 256
	return models.StatusRunning
}

// Stack operations

func (rp *RunningProgram) stackPush(val int) int {
	if len(rp.Stack) >= MaxStackSize {
		return models.StatusDiedStackOverflow
	}
	rp.Stack = append(rp.Stack, val)
	return models.StatusRunning
}

func (rp *RunningProgram) stackPop() (int, int) {
	if len(rp.Stack) == 0 {
		return 0, models.StatusDiedStackUnderflow
	}
	val := rp.Stack[len(rp.Stack)-1]
	rp.Stack = rp.Stack[:len(rp.Stack)-1]
	return val, models.StatusRunning
}

func (rp *RunningProgram) stackPeek() (int, int) {
	if len(rp.Stack) == 0 {
		return 0, models.StatusDiedStackUnderflow
	}
	return rp.Stack[len(rp.Stack)-1], models.StatusRunning
}

// Opcode implementations

func (rp *RunningProgram) runNoop() int {
	return rp.incIP()
}

func (rp *RunningProgram) runPush() int {
	if status := rp.stackPush(rp.arg()); status != models.StatusRunning {
		return status
	}
	return rp.incIP()
}

func (rp *RunningProgram) runDrop() int {
	if _, status := rp.stackPop(); status != models.StatusRunning {
		return status
	}
	return rp.incIP()
}

func (rp *RunningProgram) runDupe() int {
	val, status := rp.stackPeek()
	if status != models.StatusRunning {
		return status
	}
	if status := rp.stackPush(val); status != models.StatusRunning {
		return status
	}
	return rp.incIP()
}

func (rp *RunningProgram) runDupe2() int {
	val, status := rp.stackPeek()
	if status != models.StatusRunning {
		return status
	}
	if status := rp.stackPush(val); status != models.StatusRunning {
		return status
	}
	if status := rp.stackPush(val); status != models.StatusRunning {
		return status
	}
	return rp.incIP()
}

func (rp *RunningProgram) runSwap() int {
	if len(rp.Stack) < 2 {
		return models.StatusDiedStackUnderflow
	}
	last := len(rp.Stack) - 1
	rp.Stack[last], rp.Stack[last-1] = rp.Stack[last-1], rp.Stack[last]
	return rp.incIP()
}

func (rp *RunningProgram) runInc() int {
	if len(rp.Stack) == 0 {
		return models.StatusDiedStackUnderflow
	}
	rp.Stack[len(rp.Stack)-1] += rp.arg()
	return rp.incIP()
}

func (rp *RunningProgram) runAdd() int {
	if len(rp.Stack) < 2 {
		return models.StatusDiedStackUnderflow
	}
	val1, _ := rp.stackPop()
	val2, _ := rp.stackPop()
	value := val1 + val2
	if value < MinArgValue || value > MaxArgValue {
		return models.StatusDiedArgOverflow
	}
	rp.Stack = append(rp.Stack, value%256)
	return rp.incIP()
}

func (rp *RunningProgram) runSubtract() int {
	if len(rp.Stack) < 2 {
		return models.StatusDiedStackUnderflow
	}
	arg1, _ := rp.stackPop()
	arg2, _ := rp.stackPop()
	value := arg2 - arg1
	if value < MinArgValue || value > MaxArgValue {
		return models.StatusDiedArgOverflow
	}
	rp.Stack = append(rp.Stack, value%256)
	return rp.incIP()
}

func (rp *RunningProgram) runDivide() int {
	if len(rp.Stack) == 0 {
		return models.StatusDiedStackUnderflow
	}
	arg2 := rp.arg()
	if arg2 != 0 {
		arg1, _ := rp.stackPop()
		value := arg1 / arg2
		if value < MinArgValue || value > MaxArgValue {
			return models.StatusDiedArgOverflow
		}
		rp.Stack = append(rp.Stack, value%256)
	}
	return rp.incIP()
}

func (rp *RunningProgram) runMultiply() int {
	if len(rp.Stack) == 0 {
		return models.StatusDiedStackUnderflow
	}
	arg2 := rp.arg()
	if arg2 != 0 {
		arg1, _ := rp.stackPop()
		value := arg1 * arg2
		if value < MinArgValue || value > MaxArgValue {
			return models.StatusDiedArgOverflow
		}
		rp.Stack = append(rp.Stack, value%256)
	}
	return rp.incIP()
}

func (rp *RunningProgram) runGT() int {
	if len(rp.Stack) == 0 {
		return models.StatusDiedStackUnderflow
	}
	val, _ := rp.stackPop()
	if val > rp.arg() {
		rp.Stack = append(rp.Stack, 1)
	} else {
		rp.Stack = append(rp.Stack, 0)
	}
	return rp.incIP()
}

func (rp *RunningProgram) runLT() int {
	if len(rp.Stack) == 0 {
		return models.StatusDiedStackUnderflow
	}
	val, _ := rp.stackPop()
	if val < rp.arg() {
		rp.Stack = append(rp.Stack, 1)
	} else {
		rp.Stack = append(rp.Stack, 0)
	}
	return rp.incIP()
}

func (rp *RunningProgram) runNegate() int {
	if len(rp.Stack) == 0 {
		return models.StatusDiedStackUnderflow
	}
	rp.Stack[len(rp.Stack)-1] *= -1
	return rp.incIP()
}

func (rp *RunningProgram) runJump() int {
	if rp.arg() == 0 {
		return models.StatusDiedJump0
	}
	rp.IP = (rp.IP + rp.arg()) % 256
	if rp.IP < 0 {
		rp.IP += 256
	}
	return models.StatusRunning
}

func (rp *RunningProgram) runJumpZ() int {
	val, status := rp.stackPop()
	if status != models.StatusRunning {
		return status
	}
	if val == 0 {
		return rp.runJump()
	}
	return rp.incIP()
}

func (rp *RunningProgram) runJumpG() int {
	val, status := rp.stackPop()
	if status != models.StatusRunning {
		return status
	}
	if val > 0 {
		return rp.runJump()
	}
	return rp.incIP()
}

func (rp *RunningProgram) runHCF(inst *MemoryCell) int {
	if inst.ProgramNum == rp.ProgramNum {
		return models.StatusDiedExecutingOwnHCF
	}
	return models.StatusDiedExecutingEnemyHCF
}

func (rp *RunningProgram) runScan() int {
	if len(rp.Stack) == 0 {
		return models.StatusDiedStackUnderflow
	}

	stackVal, _ := rp.stackPop()
	pos := (rp.IP + stackVal) % 256
	if pos < 0 {
		pos += 256
	}
	count := rp.arg()
	var instFound *MemoryCell

	if count != 0 {
		first := pos
		rp.ReadAddrFirst = &first
	}

	for count != 0 {
		instFound = rp.Memory[pos]
		if instFound != nil && instFound.Opcode != "" {
			break
		}

		if count > 0 {
			pos = (pos + 1) % 256
			count--
		} else {
			pos = (pos - 1 + 256) % 256
			count++
		}
	}

	if rp.ReadAddrFirst != nil {
		last := pos
		rp.ReadAddrLast = &last

		// Reverse first and last when scan was backwards
		if rp.arg() < 0 {
			rp.ReadAddrFirst, rp.ReadAddrLast = rp.ReadAddrLast, rp.ReadAddrFirst
		}
	}

	if instFound == nil || instFound.Opcode == "" {
		rp.Stack = append(rp.Stack, 0)
	} else if instFound.IsHCF() {
		rp.Stack = append(rp.Stack, -1)
	} else {
		rp.Stack = append(rp.Stack, 1)
	}

	return rp.incIP()
}

func (rp *RunningProgram) runCopy(indirect bool, copy2 int) int {
	if len(rp.Stack) == 0 {
		return models.StatusDiedStackUnderflow
	}

	stackVal, _ := rp.stackPop()
	writes := make([]map[string]interface{}, 0)

	unlimitedFireSpread := false
	if envVal := os.Getenv("UNLIMITED_FIRE_SPREAD"); envVal != "" {
		if val, err := strconv.Atoi(envVal); err == nil && val == 1 {
			unlimitedFireSpread = true
		}
	}

	for i := 0; i <= copy2; i++ {
		srcPos := (rp.IP + i + rp.arg()) % 256
		if srcPos < 0 {
			srcPos += 256
		}

		dstPos := (rp.IP + i + stackVal) % 256
		if dstPos < 0 {
			dstPos += 256
		}

		if indirect && rp.Memory[dstPos] != nil && rp.Memory[dstPos].Argument != nil {
			dstPos = (dstPos + *rp.Memory[dstPos].Argument) % 256
			if dstPos < 0 {
				dstPos += 256
			}
		}

		explosion := false
		if rp.Memory[srcPos] != nil && rp.Memory[srcPos].IsHCF() &&
			rp.Memory[dstPos] != nil && rp.Memory[dstPos].IsHCF() {
			explosion = true
		}

		rp.Memory[dstPos] = rp.Memory[srcPos]

		first := srcPos
		rp.ReadAddrFirst = &first
		rp.ReadAddrLast = &first

		wrote := make(map[string]interface{})
		wrote["addr"] = dstPos
		if rp.Memory[dstPos] != nil {
			wrote["opcode"] = rp.Memory[dstPos].Opcode
			if rp.Memory[dstPos].Argument != nil {
				wrote["argument"] = *rp.Memory[dstPos].Argument
			}
		}
		writes = append(writes, wrote)

		if explosion {
			// Spread fire
			dstPosBefore := (dstPos - 1 + 256) % 256
			for unlimitedFireSpread && rp.Memory[dstPosBefore] != nil && rp.Memory[dstPosBefore].IsHCF() {
				dstPosBefore = (dstPosBefore - 1 + 256) % 256
			}

			dstPosAfter := (dstPos + 1) % 256
			for unlimitedFireSpread && rp.Memory[dstPosAfter] != nil && rp.Memory[dstPosAfter].IsHCF() {
				dstPosAfter = (dstPosAfter + 1) % 256
			}

			rp.Memory[dstPosBefore] = rp.Memory[dstPos]
			rp.Memory[dstPosAfter] = rp.Memory[dstPos]

			explosion1 := make(map[string]interface{})
			for k, v := range wrote {
				explosion1[k] = v
			}
			explosion1["addr"] = dstPosBefore
			writes = append(writes, explosion1)

			explosion2 := make(map[string]interface{})
			for k, v := range wrote {
				explosion2[k] = v
			}
			explosion2["addr"] = dstPosAfter
			writes = append(writes, explosion2)
		}
	}

	rp.Wrotes = writes
	return rp.incIP()
}

func (rp *RunningProgram) runDecSkip() int {
	pos := (rp.IP + rp.arg()) % 256
	if pos < 0 {
		pos += 256
	}

	if rp.Memory[pos] == nil {
		rp.Memory[pos] = NewMemoryCellNoOpcode(-1, rp.ProgramNum)
	} else {
		rp.Memory[pos].IncArgument(-1)
		rp.Memory[pos].ProgramNum = rp.ProgramNum
	}

	wrote := make(map[string]interface{})
	wrote["addr"] = pos
	if rp.Memory[pos].Argument != nil {
		wrote["argument"] = *rp.Memory[pos].Argument
	}
	rp.Wrotes = []map[string]interface{}{wrote}

	if rp.Memory[pos].Argument != nil && *rp.Memory[pos].Argument == 0 {
		rp.incIP()
	}
	return rp.incIP()
}

func (rp *RunningProgram) runSetArg() int {
	if len(rp.Stack) == 0 {
		return models.StatusDiedStackUnderflow
	}
	value, _ := rp.stackPop()
	pos := (rp.IP + rp.arg()) % 256
	if pos < 0 {
		pos += 256
	}

	if rp.Memory[pos] == nil {
		rp.Memory[pos] = NewMemoryCellNoOpcode(value, rp.ProgramNum)
	} else {
		arg := value
		rp.Memory[pos].Argument = &arg
		rp.Memory[pos].ProgramNum = rp.ProgramNum
	}

	wrote := make(map[string]interface{})
	wrote["addr"] = pos
	wrote["argument"] = value
	rp.Wrotes = []map[string]interface{}{wrote}

	return rp.incIP()
}

func (rp *RunningProgram) runPushArg() int {
	if len(rp.Stack) == 0 {
		return models.StatusDiedStackUnderflow
	}
	stackVal, _ := rp.stackPop()
	pos := (rp.IP + stackVal) % 256
	if pos < 0 {
		pos += 256
	}

	value := 0
	if rp.Memory[pos] != nil && rp.Memory[pos].Argument != nil {
		value = *rp.Memory[pos].Argument
	}
	rp.Stack = append(rp.Stack, value)
	return rp.incIP()
}

func (rp *RunningProgram) runIncArg() int {
	if len(rp.Stack) == 0 {
		return models.StatusDiedStackUnderflow
	}
	stackVal, _ := rp.stackPop()
	pos := (rp.IP + stackVal) % 256
	if pos < 0 {
		pos += 256
	}

	if rp.Memory[pos] == nil {
		rp.Memory[pos] = NewMemoryCellNoOpcode(rp.arg(), rp.ProgramNum)
	} else {
		rp.Memory[pos].IncArgument(rp.arg())
		rp.Memory[pos].ProgramNum = rp.ProgramNum
	}

	wrote := make(map[string]interface{})
	wrote["addr"] = pos
	if rp.Memory[pos].Argument != nil {
		wrote["argument"] = *rp.Memory[pos].Argument
	}
	rp.Wrotes = []map[string]interface{}{wrote}

	return rp.incIP()
}

// StackToString converts the stack to a comma-separated string
func StackToString(stack []int) string {
	if len(stack) == 0 {
		return ""
	}
	strStack := make([]string, len(stack))
	for i, v := range stack {
		strStack[i] = strconv.Itoa(v)
	}
	return strings.Join(strStack, ",")
}
